src/mod/server.mod/servmsg.c

```diff
@@ -1134,6 +1137,147 @@ static int got311(char *from, char *msg)
   return 0;
 }
 
+static int gotcap(char *from, char *msg)
+{
+  char *cmd;
+
+  newsplit(&msg);
+  putlog(LOG_SERV, "*", "CAP: %s", msg);
+  cmd = newsplit(&msg);
+  fixcolon(msg);
+  if (!strcmp(cmd, "LS")) {
+    putlog(LOG_MISC, "*", "%s supports CAP sub-commands: %s", from, msg);
+    for (cmd = newsplit(&msg); cmd[0]; cmd = newsplit(&msg)) {
+      if (!strcmp(cmd, "sasl")) {
+        putlog(LOG_SERV, "*", "SASL: CAP request sasl");
+        dprintf(DP_MODE, "CAP REQ :sasl\n");
+      }
+    }
+  } else if (!strcmp(cmd, "ACK")) {
+    putlog(LOG_MISC, "*", "%s acknowledged %s", from, msg);
+    for (cmd = newsplit(&msg); cmd[0]; cmd = newsplit(&msg)) {
+      if (!strcmp(cmd, "sasl")) {
+        putlog(LOG_SERV, "*", "SASL: put AUTHENTICATE %s", sasl_mechanism);
+        dprintf(DP_MODE, "AUTHENTICATE %s\n", sasl_mechanism);
+      }
+    }
+  } else if (cmd[0]) {
+    putlog(LOG_MISC, "*", "%s subcommand %s unknown, CAP END", from, cmd);
+    dprintf(DP_MODE, "CAP END\n");
+  }
+
+  return 1;
+}
+
+static int gotauthenticate(char *from, char *msg)
+{
+  char src[256] = ""; // FIXME: size
+  char *s;
+  size_t slen;
+  unsigned char dst[256] = ""; // FIXME: size
+  size_t olen;
+  unsigned char dst2[256] = ""; // FIXME: size
+  unsigned int olen2;
+#ifdef HAVE_OPENSSL_SSL_H
+  FILE *fp;
+  EC_KEY *eckey;
+  EVP_PKEY *privateKey;
+#endif
+
+  putlog(LOG_SERV, "*", "SASL: got AUTHENTICATE %s", msg);
+  if (msg[0] == '+') {
+    s = src;
+    /* Don't use snprintf due to \0 inside */
+    if (!strcmp(sasl_mechanism, "PLAIN")) {
+      strcpy(s, sasl_username);
+      s += strlen(sasl_username) + 1;
+      strcpy(s, sasl_username);
+      s += strlen(sasl_username) + 1;
+      strcpy(s, sasl_password);
+      s += strlen(sasl_password);
+    }
+    else if (!strcmp(sasl_mechanism, "ECDSA-NIST256P-CHALLENGE")) {
+      strcpy(s, sasl_username);
+      s += strlen(sasl_username) + 1;
+      strcpy(s, sasl_username);
+      s += strlen(sasl_username);
+    }
+    else /* sasl_mechanism == EXTERNAL */
+      *s++ = '+';
```

Comment:
This isn't valid a response for EXTERNAL (unless the server has a user account named "+"). The response needs to either contain the username (like you're doing for PLAIN), or be an empty string.

An empty string only becomes a `+` when it's encoded and sent to the server via AUTHENTICATE (and vice versa, when a `+` is received, it decodes to an empty string).

Reply:
Thank you! Fixed with commit 6c48bed8365eeda333b78e106c6231593bf2a9c7.